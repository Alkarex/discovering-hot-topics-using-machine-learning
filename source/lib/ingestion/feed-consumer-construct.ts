#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/


import { Construct, Duration } from '@aws-cdk/core';
import { Runtime, StartingPosition, Code, IFunction } from '@aws-cdk/aws-lambda';
import { StreamEncryption, Stream, IStream } from '@aws-cdk/aws-kinesis';
import { buildLambdaFunction } from '@aws-solutions-constructs/core';
import { KinesisEventSource } from '@aws-cdk/aws-lambda-event-sources';


export interface FeedConsumerProps {
    readonly environment?: { [key: string]: string }
    readonly timeout: Duration,
    readonly runtime: Runtime,
    readonly code: Code,
    readonly batchSize?: number,
    readonly retryAttempts?: number,
    readonly shardCount?: number,
    readonly retentionPeriod?: Duration
}

export class FeedConsumer extends Construct {

    private readonly feedConsumerStream: IStream;
    private consumerFunction: IFunction;

    constructor(scope: Construct, id: string, props: FeedConsumerProps) {
        super(scope, id);

        this.feedConsumerStream = new Stream (this, 'InferenceStreamToWF', {
            encryption: StreamEncryption.MANAGED,
            retentionPeriod: props.retentionPeriod,
            shardCount: props.shardCount
        });

        this.consumerFunction = buildLambdaFunction(this, {
            lambdaFunctionProps: {
                runtime: props.runtime,
                handler: 'index.handler',
                code: props.code,
                timeout: Duration.minutes(5),
                environment: props.environment,
                events: [ new KinesisEventSource(this.feedConsumerStream, {
                        startingPosition: StartingPosition.TRIM_HORIZON,
                        batchSize: props.batchSize,
                        retryAttempts: props.retryAttempts
                    }
                )]
            }
        });
    }

    public get lambdaFunction(): IFunction {
        return this.consumerFunction;
    }

    public get kinesisStream(): IStream {
        return this.feedConsumerStream;
    }
}
