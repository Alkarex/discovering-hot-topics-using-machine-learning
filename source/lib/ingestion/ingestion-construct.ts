#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/


import { Construct, Duration, Aws } from '@aws-cdk/core';
import { Runtime, Code } from '@aws-cdk/aws-lambda';
import { FeedProducer } from './feed-producer-construct';
import { FeedConsumer } from './feed-consumer-construct';
import { PolicyStatement, Effect } from '@aws-cdk/aws-iam';

export interface IngestionProps {
    readonly stateMachineArn: string,
    readonly solutionName: string,
    readonly ingestFrequency: string,
    readonly queryParameter: string,
    readonly supportedLang: string //comma separated values of language codes
    readonly credentialKeyPath: string
}

export class Ingestion extends Construct {
    constructor(scope: Construct, id: string, props: IngestionProps) {
        super(scope, id);

        // start creation of Kinesis consumer that invokes state machine
        const invokeStepFunctionPolicy = new PolicyStatement({
            effect: Effect.ALLOW,
            resources: [ props.stateMachineArn ],
            actions: ['states:StartExecution']
        });

        const feedConsumerlambda = new FeedConsumer(this, 'FeedConsumer', {
            environment: {
                WORKFLOW_ARN: props.stateMachineArn,
            },
            runtime: Runtime.NODEJS_12_X,
            code: Code.fromAsset(`${__dirname}/../../lambda/ingestion-consumer`),
            timeout: Duration.minutes(5)
        });

        feedConsumerlambda.lambdaFunction.addToRolePolicy(invokeStepFunctionPolicy);
        // end creation of Kinesis consumer that invokes state machine


        // start creation of Lambda Kinesis producer that fetches social media feed
        new FeedProducer(this, 'FeedProducerConstruct', {
            timeout: Duration.minutes(5),
            stream: feedConsumerlambda.kinesisStream,
            runtime: Runtime.NODEJS_12_X,
            code: Code.fromAsset(`${__dirname}/../../lambda/ingestion-producer`),
            solutionName: props.solutionName,
            supportedLang: props.supportedLang,
            ingestFrequency: props.ingestFrequency,
            queryParameter: props.queryParameter,
            credentialKeyPath: props.credentialKeyPath
        });
        // end creation of Lambda Kinesis producer that fetches social media feed
    }
}
