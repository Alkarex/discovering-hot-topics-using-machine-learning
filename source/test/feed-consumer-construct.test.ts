/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Runtime, Code } from "@aws-cdk/aws-lambda";
import { SynthUtils } from '@aws-cdk/assert';
import { Duration, Stack } from '@aws-cdk/core';
import { FeedConsumer } from '../lib/ingestion/feed-consumer-construct';
import '@aws-cdk/assert/jest';

test('Test Lambda with Kinesis', () => {
    const stack = new Stack();

    new FeedConsumer(stack,'FeedConsumerConstruct', {
        environment: {
            WORKFLOW_ARN: 'arn:aws:states:us-east-1:someaccountid:stateMachine:WorkflowEngine12346891012-ad234ab'
        },
        timeout: Duration.minutes(5),
        runtime: Runtime.NODEJS_12_X,
        code: Code.fromAsset(`${__dirname}/../lambda/ingestion-producer`),
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();

});