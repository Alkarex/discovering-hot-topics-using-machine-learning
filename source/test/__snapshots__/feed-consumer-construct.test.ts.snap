// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Lambda with Kinesis 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters73a6625c70eaedf02bd1901a58569aa68070c8165d50bcb4e71a68a5c2f1a3c9ArtifactHashAECF6781": Object {
      "Description": "Artifact hash for asset \\"73a6625c70eaedf02bd1901a58569aa68070c8165d50bcb4e71a68a5c2f1a3c9\\"",
      "Type": "String",
    },
    "AssetParameters73a6625c70eaedf02bd1901a58569aa68070c8165d50bcb4e71a68a5c2f1a3c9S3Bucket5D8B64E4": Object {
      "Description": "S3 bucket for asset \\"73a6625c70eaedf02bd1901a58569aa68070c8165d50bcb4e71a68a5c2f1a3c9\\"",
      "Type": "String",
    },
    "AssetParameters73a6625c70eaedf02bd1901a58569aa68070c8165d50bcb4e71a68a5c2f1a3c9S3VersionKeyB1EF4AC7": Object {
      "Description": "S3 key for asset version \\"73a6625c70eaedf02bd1901a58569aa68070c8165d50bcb4e71a68a5c2f1a3c9\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "FeedConsumerConstructInferenceStreamToWF60D525F1": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "EncryptionType": "KMS",
          "KeyId": "alias/aws/kinesis",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "FeedConsumerConstructLambdaFunction5D2FC9AE": Object {
      "DependsOn": Array [
        "FeedConsumerConstructLambdaFunctionServiceRoleDefaultPolicyE9FC8BA4",
        "FeedConsumerConstructLambdaFunctionServiceRole1C4C9124",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "Lambda functions has the required permission to write CloudWatch Logs. It uses custom policy instead of arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole with more tighter permissions.",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters73a6625c70eaedf02bd1901a58569aa68070c8165d50bcb4e71a68a5c2f1a3c9S3Bucket5D8B64E4",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters73a6625c70eaedf02bd1901a58569aa68070c8165d50bcb4e71a68a5c2f1a3c9S3VersionKeyB1EF4AC7",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters73a6625c70eaedf02bd1901a58569aa68070c8165d50bcb4e71a68a5c2f1a3c9S3VersionKeyB1EF4AC7",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "WORKFLOW_ARN": "arn:aws:states:us-east-1:someaccountid:stateMachine:WorkflowEngine12346891012-ad234ab",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "FeedConsumerConstructLambdaFunctionServiceRole1C4C9124",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "FeedConsumerConstructLambdaFunctionKinesisEventSourceFeedConsumerConstructInferenceStreamToWFFDB15F166737E8CE": Object {
      "Properties": Object {
        "BatchSize": 100,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "FeedConsumerConstructInferenceStreamToWF60D525F1",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "FeedConsumerConstructLambdaFunction5D2FC9AE",
        },
        "StartingPosition": "TRIM_HORIZON",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "FeedConsumerConstructLambdaFunctionServiceRole1C4C9124": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:aws:logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaFunctionServiceRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "FeedConsumerConstructLambdaFunctionServiceRoleDefaultPolicyE9FC8BA4": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "FeedConsumerConstructInferenceStreamToWF60D525F1",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "kinesis:DescribeStream",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "FeedConsumerConstructInferenceStreamToWF60D525F1",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "FeedConsumerConstructLambdaFunctionServiceRoleDefaultPolicyE9FC8BA4",
        "Roles": Array [
          Object {
            "Ref": "FeedConsumerConstructLambdaFunctionServiceRole1C4C9124",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
